### PR TITLE
Fix segfault on return type

### PR DESCRIPTION
**Pull request by wsmoses**
Fixes https://github.com/EnzymeAD/Enzyme.jl/issues/2074

### CODE REVIEW

File: lib/EnzymeCore/src/rules.jl [hunk 1]
```diff
@@ -208,18 +208,36 @@ function isapplicable(@nospecialize(f), @nospecialize(TT);
...
+        if caller isa Core.MethodInstance
+            add_mt_backedge!(caller, mt, sig)
+        elseif caller isa Core.Compiler.MethodLookupResult
+            for j = 1:Core.Compiler.length(caller)
+                cmatch = Core.Compiler.getindex(caller, j)::Core.MethodMatch
+                cspec = Core.Compiler.specialize_method(cmatch)::Core.MethodInstance
+                add_mt_backedge!(cspec, mt, sig)
+            end
+        end
     end
```
Comment (vchuravy) [inline lib/EnzymeCore/src/rules.jl:221]: Can you explain why this code became necessary?

Comment (wsmoses) [reply lib/EnzymeCore/src/rules.jl:221]: Yeah so basically it was required to fix a bug in the rules tests.

in particular if a function was defined first without a rule, absint would succeed and let’s say inline the function. The only other time we add a back edge is when we take the actual generated llvm functions, which only exist for non Julia inlined code. So we would never add a back edge for the rule currently without a rule.

we would then go and add a rule, but that wouldn’t cause an invalidation and thus end up with the wrong code

File: src/compiler/interpreter.jl [hunk 2]
```diff
@@ -214,21 +214,27 @@ function Core.Compiler.abstract_call_gf_by_type(
     callinfo = ret.info
     method_table = Core.Compiler.method_table(interp)
     specTypes = simplify_kw(atype)
+    caller = if callinfo isa Core.Compiler.MethodMatchInfo && callinfo.results isa Core.Compiler.MethodLookupResult
```
Comment (vchuravy) [inline src/compiler/interpreter.jl:217]: This seems wrong. We want to recompile the method one-above in the call-graph not the primal method from whence we came?

Comment (wsmoses) [reply src/compiler/interpreter.jl:217]: I don’t follow? If a rule is defined for f we want in invalidation edge exist from inactive(f) -> f, right?

Comment (wsmoses) [reply src/compiler/interpreter.jl:217]: For example if f defines a custom rule, we need to invalidate autodiff(f) to now just call that rule

File: lib/EnzymeCore/src/rules.jl (hunk 1 above)
Comment (vchuravy) [reply lib/EnzymeCore/src/rules.jl:221]: Yeah, that was known as a bug...

Julia solves this, by having a new method definition invalidate the method-table. We could change our rule definitions to a macro form that does that.

**Pull request merged**